mkosi/__init__.py

```diff
@@ -3398,7 +3476,133 @@ def boot_directory(state: MkosiState, kver: str) -> Path:
+    return state.workspace / f"{kver}/initrd"
+
+
+def init_symlink_path(state: MkosiState, kver: str) -> Path:
+    return initrd_dir_path(state, kver) / "init"
+
+
+def initrd_release_path(state: MkosiState, kver: str) -> Path:
+    return initrd_dir_path(state, kver) / "etc/initrd-release"
+
+
+def make_init_symlink(state: MkosiState, kver: str) -> Path:
+    init = init_symlink_path(state, kver)
+    init.unlink(missing_ok=True)
+
+    if not init.parent.exists():
+        init.parent.mkdir(parents=True)
+
+    init.symlink_to('usr/lib/systemd/systemd')
+    return init
+
+
+def write_initrd_release(state: MkosiState, kver: str) -> Path:
+    initrd_release = initrd_release_path(state, kver)
+    initrd_release.unlink(missing_ok=True)
+
+    if not initrd_release.parent.exists():
+        initrd_release.parent.mkdir(parents=True)
+
+    with initrd_release.open('wt') as out:
+        os_release = read_os_release(state.root)
+
+        # Replacing fields in the original dictionary should maintain the order
+        update_suffixed(os_release, 'NAME', 'Linux')
+        update_suffixed(os_release, 'PRETTY_NAME', 'Linux')
+        os_release['VARIANT'] = 'mkosi-initrd'
+        os_release['VARIANT_ID'] = 'mkosi-initrd'
+
+        for name, value in os_release.items():
+            if value:
+                print(f'{name}={value!r}', file=out)
+
+    return initrd_release
+
+
+def generate_kernel_modules_initrd(state: MkosiState) -> None:
+    with complete_step("Generating initramfs images for kernel modules…"):
+        kernel_modules_in_initrds = list_kernel_modules_in_initrds(state)
+        for kver, _ in gen_kernel_images(state):
+            cpio_path = kernel_modules_initrd_cpio_path(state, kver)
+            assert cpio_path
+
+            # Kernel modules
+            if kver not in kernel_modules_in_initrds:
+                make_cpio(state.root, find_files(state.root / "usr/lib/modules", state.root), cpio_path)
+            else:
+                MkosiPrinter.info(f"Kernel modules for {kver} found in {kernel_modules_in_initrds[kver]}: skipping")
+
+            # /init, /etc/initrd-release
+            initrd_writers = {
+                "initrd-release": write_initrd_release,
+                "init": make_init_symlink,
+            }
+            initrd_files = []
+            for file in initrd_writers:
+                if file not in kernel_modules_in_initrds:
+                    path = initrd_writers[file](state, kver)
+                    initrd_files.append(path.relative_to(initrd_dir_path(state, kver)))
+                else:
+                    MkosiPrinter.info(f"{file} found in {kernel_modules_in_initrds[file]}: skipping")
+            if initrd_files:
+                if cpio_path.exists():
+                    update_cpio(initrd_dir_path(state, kver), iter(initrd_files), cpio_path)
+                else:
+                    make_cpio(initrd_dir_path(state, kver), iter(initrd_files), cpio_path)
```

Comment:
So after thinking this over a little more, I think we should do this differently. Instead of adding these files in the kernel modules initrd, we should add them when we build the initial initrd cpio.

So instead of adding these here, when the output format is a cpio, we should add the symlink and /etc/initrd-release to the image before packing it up as a cpio.

Reply:
> when we build the initial initrd cpio.

You mean the initrd that is provided by the user via the `--initrd` option?

Reply:
Exactly, that initrd should have the init symlink and /etc/initrd-release, and we can easily add those when generating the initrd.